src/EnergyPlus/PlantLoopHeatPumpEIR.cc

```diff
@@ -1292,4 +1299,1344 @@ void EIRPlantLoopHeatPump::oneTimeInit(EnergyPlusData &state)
+                  EIRPlantLoopHeatPumps::EIRFuelFiredHeatPump::subtract},
+    };
+
+    bool errorsFound = false;
+    std::string &cCurrentModuleObject = state.dataIPShortCut->cCurrentModuleObject;
+    for (auto &classToInput : classesToInput) {
+        cCurrentModuleObject = DataPlant::PlantEquipTypeNames[static_cast<int>(classToInput.thisType)];
+
+        DataLoopNode::ConnectionObjectType objType = static_cast<DataLoopNode::ConnectionObjectType>(
+            getEnumerationValue(BranchNodeConnections::ConnectionObjectTypeNamesUC, UtilityRoutines::MakeUPPERCase(cCurrentModuleObject)));
+
+        auto const instances = state.dataInputProcessing->inputProcessor->epJSON.find(cCurrentModuleObject);
+        if (instances == state.dataInputProcessing->inputProcessor->epJSON.end()) continue;
+        auto &instancesValue = instances.value();
+        for (auto instance = instancesValue.begin(); instance != instancesValue.end(); ++instance) {
+            auto const &fields = instance.value();
+            auto const &thisObjectName = instance.key();
+            state.dataInputProcessing->inputProcessor->markObjectAsUsed(cCurrentModuleObject, thisObjectName);
+
+            EIRFuelFiredHeatPump thisPLHP;
+
+            thisPLHP.EIRHPType = classToInput.thisType;
+            std::string companionCoilFieldTag = "companion_heating_heat_pump_name";
+            std::string refCapFieldTag = "nominal_cooling_capacity";
+            if (thisPLHP.EIRHPType == DataPlant::PlantEquipmentType::HeatPumpFuelFiredHeating) {
+                companionCoilFieldTag = "companion_cooling_heat_pump_name";
+                refCapFieldTag = "nominal_heating_capacity";
+            }
+
+            // A1-A3
+            thisPLHP.name = UtilityRoutines::MakeUPPERCase(thisObjectName);
+            std::string loadSideInletNodeName = UtilityRoutines::MakeUPPERCase(fields.at("water_inlet_node_name").get<std::string>());
+            std::string loadSideOutletNodeName = UtilityRoutines::MakeUPPERCase(fields.at("water_outlet_node_name").get<std::string>());
+            // Implicit
+            // std::string condenserType = "AIRSOURCE"; // UtilityRoutines::MakeUPPERCase(fields.at("condenser_type").get<std::string>());
+            thisPLHP.waterSource = false;
+            thisPLHP.airSource = false;
+            // A4
+            std::string sourceSideInletNodeName = UtilityRoutines::MakeUPPERCase(fields.at("air_source_node_name").get<std::string>());
+            std::string sourceSideOutletNodeName =
+                "DUMMY_CONDENSER"; // UtilityRoutines::MakeUPPERCase(fields.at("source_side_outlet_node_name").get<std::string>());
+            srand(time(NULL));
+            for (int i = 1; i <= 3; ++i) {
+                sourceSideOutletNodeName += ("_" + std::to_string(rand()));
+            }
+
+            // A5
+            if (fields.find(companionCoilFieldTag) != fields.end()) { // optional field
```

Comment:
I saw that there was a discussion about this elsewhere so I will use this opportunity to comment on this idiom. By doing an `at("x")` after a `find("x")` you are searching a binary tree (a `map` is a binary tree, which is why I keep telling people not to use it unnecessarily) for the same string twice despite the fact that by getting to the `at` you know you've already found it. A better idiom here is:

```cpp
auto foundX = fields.find("x"); // fine to use plain `auto` for an iterator
if (foundX != fields.end()) 
   thisPLHP.companionCoilName = MakeUPPERCase(foundX.value().get<std::string>());
```

@rraustad, this idiom is all over `UnitarySystem.cc` fwiw.

Reply:
This works fine now! One question is that when doing this pattern, does the scenario of `foundX == fields.end()` need to be handled---e.g. in an `else {}` block?

Reply:
That's context-dependent. If you have to explicitly do something if the field is not given then that would go on in the `else {}` block.